dask_match/expr/core.py

```diff
@@ -0,0 +1,578 @@
+
+    _parameters = ["frame", "dtypes"]
+    operation = M.astype
+
+
+class Apply(Elemwise):
+    """A good example of writing a less-trivial blockwise operation"""
+
+    _parameters = ["frame", "function", "args", "kwargs"]
+    _defaults = {"args": (), "kwargs": {}}
+    operation = M.apply
+
+    @property
+    def _meta(self):
+        return self.frame._meta.apply(self.function, *self.args, **self.kwargs)
+
+    def _layer(self):
+        return {
+            (self._name, i): (
+                apply,
+                M.apply,
+                [(self.frame._name, i), self.function] + list(self.args),
+                self.kwargs,
+            )
+            for i in range(self.npartitions)
+        }
+
+
+class Filter(Blockwise):
+    _parameters = ["frame", "predicate"]
+    operation = operator.getitem
+
+    @classmethod
+    def _replacement_rules(self):
+        df = Wildcard.dot("df")
+        condition = Wildcard.dot("condition")
+        columns = Wildcard.dot("columns")
+
+        # Project columns down through dataframe
+        # df[df.x > 1].y -> df.y[df.x > 1]
+        yield ReplacementRule(
+            Pattern(Filter(df, condition)[columns]),
+            lambda df, condition, columns: df[columns][condition],
+        )
+
+
+class Projection(Elemwise):
+    """Column Selection"""
+
+    _parameters = ["frame", "columns"]
+    operation = operator.getitem
+
+    def _divisions(self):
+        return self.frame.divisions
+
+    @property
+    def columns(self):
+        return self.operand("columns")
+
+    @property
+    def _meta(self):
+        return self.frame._meta[self.columns]
+
+    def _layer(self):
+        return {
+            (self._name, i): (operator.getitem, (self.frame._name, i), self.columns)
+            for i in range(self.npartitions)
+        }
+
+    def __str__(self):
+        base = str(self.frame)
+        if " " in base:
+            base = "(" + base + ")"
+        return f"{base}[{repr(self.columns)}]"
+
+
+class ProjectIndex(Elemwise):
+    """Column Selection"""
+
+    _parameters = ["frame"]
+    operation = getattr
+
+    def _divisions(self):
+        return self.frame.divisions
+
+    @property
+    def _meta(self):
+        return self.frame._meta.index
+
+    def _layer(self):
+        return {
+            (self._name, i): (getattr, (self.frame._name, i), "index")
+            for i in range(self.npartitions)
+        }
+
+
+class Binop(Elemwise):
+    _parameters = ["left", "right"]
+    arity = Arity.binary
+
+    def _layer(self):
+        return {
+            (self._name, i): (
+                self.operation,
+                (self.left._name, i) if isinstance(self.left, Expr) else self.left,
+                (self.right._name, i) if isinstance(self.right, Expr) else self.right,
+            )
+            for i in range(self.npartitions)
+        }
+
+    def __str__(self):
+        return f"{self.operand('left')} {self._operator_repr} {self.operand('right')}"
```

Comment:
My preference would be to leave these as `self.left` and `self.right`.  This is subjective though and I'm open to a larger discussion on this topic.  I suspect that others may agree with you.